internal/controllers/phase_ensure_catalogsource.go

```diff
@@ -0,0 +1,167 @@
+	k8sApiErrors "k8s.io/apimachinery/pkg/api/errors"
+	"k8s.io/apimachinery/pkg/api/meta"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller/controllerutil"
+
+	addonsv1alpha1 "github.com/openshift/addon-operator/apis/addons/v1alpha1"
+)
+
+const catalogSourcePublisher = "OSD Red Hat Addons"
+
+// Ensure existence of the CatalogSource specified in the given addon resource
+// returns a bool that signals the caller to stop reconciliation and retry later
+func (r *AddonReconciler) ensureCatalogSource(
+	ctx context.Context, log logr.Logger, addon *addonsv1alpha1.Addon) (stop, retry bool, err error) {
+	var targetNamespace, catalogSourceImage string
+
+	switch addon.Spec.Install.Type {
+	case addonsv1alpha1.OwnNamespace:
+		if addon.Spec.Install.OwnNamespace == nil ||
+			len(addon.Spec.Install.OwnNamespace.Namespace) == 0 {
+			// invalid/missing configuration
+			// TODO: Move error reporting into webhook and reduce this code to a sanity check.
+			return true, false, r.reportConfigurationError(ctx, addon, ".spec.install.ownNamespace.namespace is required when .spec.install.type = OwnNamespace")
+		}
+		targetNamespace = addon.Spec.Install.OwnNamespace.Namespace
+		if len(addon.Spec.Install.OwnNamespace.CatalogSourceImage) == 0 {
+			// invalid/missing configuration
+			// TODO: Move error reporting into webhook and reduce this code to a sanity check.
+			return true, false, r.reportConfigurationError(ctx, addon, ".spec.install.ownNamespacee.catalogSourceImage is required when .spec.install.type = OwnNamespace")
+		}
+		catalogSourceImage = addon.Spec.Install.OwnNamespace.CatalogSourceImage
+
+	case addonsv1alpha1.AllNamespaces:
+		if addon.Spec.Install.AllNamespaces == nil ||
+			len(addon.Spec.Install.AllNamespaces.Namespace) == 0 {
+			// invalid/missing configuration
+			// TODO: Move error reporting into webhook and reduce this code to a sanity check.
+			return true, false, r.reportConfigurationError(ctx, addon, ".spec.install.allNamespaces.namespace is required when .spec.install.type = AllNamespaces")
+		}
+		targetNamespace = addon.Spec.Install.AllNamespaces.Namespace
+		if len(addon.Spec.Install.AllNamespaces.CatalogSourceImage) == 0 {
+			// invalid/missing configuration
+			// TODO: Move error reporting into webhook and reduce this code to a sanity check.
+			return true, false, r.reportConfigurationError(ctx, addon, ".spec.install.allNamespaces.catalogSourceImage is required when .spec.install.type = AllNamespaces")
+		}
+		catalogSourceImage = addon.Spec.Install.AllNamespaces.CatalogSourceImage
+
+	default:
+		// Unsupported Install Type
+		// This should never happen, unless the schema validation is wrong.
+		// The .install.type property is set to only allow known enum values.
+		log.Error(fmt.Errorf("invalid Addon install type: %q", addon.Spec.Install.Type), "stopping Addon reconcilation")
+		return true, false, nil
+	}
```

Comment:
Can't we unify this block with the one in `phase_ensure_operator_group.go`?

I think such a signature should work:
```go
targetNamespace, catalogSourceImage, stop, err := r.getInstallConfig(ctx, log, addon)
if err != nil || stop {
  return true, false, err
}
```

Reply:
done!